File: QbertPython/QbertPackage/QbertBeing.py
'''
This class represents A Qbert being such as the player or enemies.

@author: Dario Urdapilleta
@version 1.0
@since: 13 nov. 2022
'''
import pygame
from enum import Enum
import pygame.math as Math

from QbertPackage import SpriteAnimation

class State(Enum):      # Enum with all the animation states
    IDDLE = 0
    JUMPING_FRONT_UP_RIGHT = 1
    JUMPING_FRONT_UP_LEFT = 2
    JUMPING_FRONT_DOWN_RIGHT = 3
    JUMPING_FRONT_DOWN_LEFT = 4
    JUMPING_BACK_UP_LEFT = 5
    JUMPING_BACK_UP_RIGHT = 6
    JUMPING_BACK_DOWN_LEFT = 7
    JUMPING_BACK_DOWN_RIGHT = 8

class QbertBeingClass(object):
    '''
    Variables:
    maxHeight: The maximum jumping height
    deltaHeight: The height change rate
    deltaMovement: The movement animation frame rate
    deltaJump: The jump animation frame rate
    position: The being's position
    texture: The being's sprite list
    movementSlice: The direction it is moving
    jumpMovement: The actual placement of the being
    animations: The animations
    state: The being's current state
    stateTime: Time for the key frames
    height: The jump height
    isMoving: True if the being is moving
    landed: True if it has landed
    isFalling: True if it's outside of the board
    pleaseRespawn: True if it needs to respawn
    '''
    maxHeight = 100         # The maximum jumping height
    deltaHeight = 4         # The height change rate
    deltaMovement = 0.02    # The movement animation frame rate
    deltaJump = 0.1         # The jump animation frame rate

    def __init__(self, texture, start, animationLength):
        '''
        Creates a new instance of a QbertBeing given the spritesheet information.
        @param self The current object
        @param fileName The Spritesheet's file name.
        @param start The start spritesheet position.
        @param animationLength The length of each animation.
        @return A new instance of the QbertBeing
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        self.state = State.IDDLE                                                            # Set the start state to IDDLE
        self.isMoving = False                                                               # It starts not moving
        self.landed = True                                                                  # It starts on the ground
        self.isFalling = False                                                              # It starts in the board
        self.pleaseRespawn = False                                                          # It doesn't need to be respawned
        self.movementSlice = Math.Vector2(0, 0)                                             # It's not moving
        self.jumpMovement = Math.Vector2(0, 0)                                              # It's not moving
        self.texture = texture.convert_alpha()                                              # Loads the being's texture
        self.animations = [None] * 5                                                        # Create the animations array
        for animationCounter in range(len(self.animations)):                                # Loop through the animations
            self.animations[animationCounter] = SpriteAnimation.SpriteAnimation(self.texture, animationLength, 60, 60, start + animationCounter)      # Create the animation
            self.animations[animationCounter].isLooping = True                              # Set it to loop
            self.animations[animationCounter].setFramesPersecond(8)                         # Set the frames per second
        self.stateTime = 0                                                                  # Reset the state time
        self.height = 0                                                                     # Set the height as 0
        
    def needsRespawn(self):
        '''
        Returns true if the being needs to respawn.
        @param self The current object
        @return True if it needs to be respwned.
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        return self.pleaseRespawn       # Please respawn this being
        
    def isStanding(self):
        '''
        Returns true if the being is standing.
        @param self The current object
        @return True if the being is standing.
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        return self.state == State.IDDLE   # Make sure the position is not null, and the state is IDDLE
    
    def isMoving(self):
        '''
        Returns true if the being is moving.
        @param self The current object
        @return True if the being is moving.
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        return self.isMoving       # Return if the being is moving
    
    def landed(self):
        '''
        Returns true if the being has landed.
        @param self The current object
        @return True if the being has landed.
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        return self.landed         # Return true if the bieng has landed
    
    def isFalling(self):
        '''
        Returns true if the being is falling.
        @param self The current object
        @return True if the being is falling.
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        return self.isFalling      # Return true if the being is falling
    
    def setHeight(self, height):
        '''
        Sets the being's height
        @param self The current object
        @param height The being's height.
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        self.height = height       # Set the height
        
    def getJumpMovement(self):
        '''
        Returns the being's actual movement
        @param self The current object
        @return The being's actual movement
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        return self.jumpMovement       # return the actual movement
    
    def getPosition(self):
        '''
        Returns the being's position as a Vector2
        @param self The current object
        @return The being's position as a Vector2
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        return self.position       # Return the position
    
    def setPosition(self, position):
        '''
        Sets the being's position
        @param self The current object
        @param position The being's new position.
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        self.position = position   # Set the position
        
    def move(self, direction):
        '''
        Moves the being torwards a direction.
        @param self The current object
        @param direction The direction the being is moving.
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        self.isMoving = True                                        # Set isMoving to true
        self.landed = False                                         # Set landed to false
        if direction == 0:                                          # north west
            self.movementSlice.y = -QbertBeingClass.deltaMovement        # Move negative on y
            self.movementSlice.x = 0                                # Leave x unchanged
            self.state = State.JUMPING_BACK_UP_LEFT                 # Change state to JUMPING_BACK_UP_LEFT
        elif direction == 1:                                        # north east
            self.movementSlice.x = -QbertBeingClass.deltaMovement        # Move negative on x
            self.movementSlice.y = 0                                # Leave y unchanged
            self.state = State.JUMPING_BACK_UP_RIGHT                # Change state to JUMPING_BACK_UP_RIGHT
        elif direction == 2:                                        # north east
            self.movementSlice.y = QbertBeingClass.deltaMovement         # Move positive on y
            self.movementSlice.x = 0                                # Leave x unchanged
            self.state = State.JUMPING_FRONT_UP_RIGHT               # Change state to JUMPING_FRONT_UP_RIGHT
        else:
            self.movementSlice.x = QbertBeingClass.deltaMovement         # Move positive on x
            self.movementSlice.y = 0                                # Leave y unchanged
            self.state = State.JUMPING_FRONT_UP_LEFT                # Change state to JUMPING_FRONT_UP_LEFT
        self.stateTime = 0                                          # Reset the state time
        
    def update(self, gameTime):
        '''
        Updates the being
        @param self The current object
        @param gameTime The game time transcurred since last update.
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        self.stateTime = self.stateTime + gameTime                  # Increase the animation state
        if self.state == State.IDDLE:                               # When it is IDDLE
            self.height = 0                                         # Set the height to 0
        elif self.state == State.JUMPING_FRONT_UP_RIGHT:            # When it is JUMPING_FRONT_UP_RIGHT
            self.height = self.height + QbertBeingClass.deltaHeight      # Increase the height
        elif self.state == State.JUMPING_FRONT_UP_LEFT:             # When it is JUMPING_FRONT_UP_LEFT
            self.height = self.height + QbertBeingClass.deltaHeight      # Increase the height
        elif self.state == State.JUMPING_FRONT_DOWN_RIGHT:          # When it is JUMPING_FRONT_DOWN_RIGHT
            self.height = self.height - QbertBeingClass.deltaHeight      # Decrease the height
        elif self.state == State.JUMPING_FRONT_DOWN_LEFT:           # When it is JUMPING_FRONT_DOWN_LEFT
            self.height = self.height - QbertBeingClass.deltaHeight      # Decrease the height
        elif self.state == State.JUMPING_BACK_UP_LEFT:              # When it is JUMPING_BACK_UP_LEFT
            self.height = self.height + QbertBeingClass.deltaHeight      # Increase the height
        elif self.state == State.JUMPING_BACK_UP_RIGHT:             # When it is JUMPING_BACK_UP_RIGHT
            self.height = self.height + QbertBeingClass.deltaHeight      # Increase the height
        elif self.state == State.JUMPING_BACK_DOWN_LEFT:            # When it is JUMPING_BACK_DOWN_LEFT
            self.height = self.height - QbertBeingClass.deltaHeight      # Decrease the height
        elif self.state == State.JUMPING_BACK_DOWN_RIGHT:           # When it is JUMPING_BACK_DOWN_RIGHT
            self.height = self.height - QbertBeingClass.deltaHeight      # Decrease the height
        else:
            self.height = 0                                         # Set the height to 0
        if self.height >= QbertBeingClass.maxHeight:                     # If the being has reached the max height
            if self.state == State.JUMPING_FRONT_UP_RIGHT:          # When it is JUMPING_FRONT_UP_RIGHT
                self.state = State.JUMPING_FRONT_DOWN_RIGHT         # Change to JUMPING_FRONT_DOWN_RIGHT
            elif self.state == State.JUMPING_FRONT_UP_LEFT:         # When it is JUMPING_FRONT_UP_LEFT
                self.state = State.JUMPING_FRONT_DOWN_LEFT          # Change to JUMPING_FRONT_DOWN_LEFT
            elif self.state == State.JUMPING_BACK_UP_LEFT:          # When it is JUMPING_BACK_UP_LEFT
                self.state = State.JUMPING_BACK_DOWN_LEFT           # Change to JUMPING_BACK_DOWN_LEFT
            elif self.state == State.JUMPING_BACK_UP_RIGHT:         # When it is JUMPING_BACK_UP_RIGHT
                self.state = State.JUMPING_BACK_DOWN_RIGHT          # Change to JUMPING_BACK_DOWN_RIGHT
        self.jumpMovement.x = self.jumpMovement.x + self.movementSlice.x    # Move the being on the x axis
        self.jumpMovement.y = self.jumpMovement.y + self.movementSlice.y    # Move the being on the y axis
        if not self.isFalling and self.height < 0:                  # Check if it's not falling and it has landed
            if self.jumpMovement.x >= 1:                            # Check the movement on x is more than 1
                self.position.x = self.position.x + 1               # Increase the position in x
            if self.jumpMovement.x <= -1:                           # Check the movement on x is lees than -1
                self.position.x = self.position.x - 1               # Decrease the position in x
            if self.jumpMovement.y >= 1:                            # Check the movement on y is more than 1
                self.position.y = self.position.y + 1               # Increase the position in y
            if self.jumpMovement.y <= -1:                           # Check the movement on y is lees than -1
                self.position.y = self.position.y - 1               # Decrease the position in y
            self.landed = True                                      # Land the being
        if self.height <= -300:                                     # Check if the height has gone less than -300
            self.pleaseRespawn = True                               # Request to respawn
        for animationCounter in range(len(self.animations)):        # Loop through the animations
            self.animations[animationCounter].update(gameTime)      # Update the animation
            
    def draw(self, screen, graphicPosition):
        '''
        Draws the being
        @param self The current object
        @param screen The screen.
        @param graphicPosition The being's acutal position
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        for animationCounter in range(len(self.animations)):                # Loop through the animations
            self.animations[animationCounter].position = Math.Vector2(graphicPosition.x, graphicPosition.y - self.height)    # Update the animations position
        if self.state == State.IDDLE:                               # When it is IDDLE
            self.animations[0].draw(screen, self.animations[0].position)        # Show the idle animation
        elif self.state == State.JUMPING_FRONT_UP_RIGHT:            # When it is JUMPING_FRONT_UP_RIGHT
            self.animations[1].draw(screen, self.animations[0].position)        # Show the JUMPING_FRONT_UP_RIGHT animation
        elif self.state == State.JUMPING_FRONT_UP_LEFT:             # When it is JUMPING_FRONT_UP_LEFT
            self.animations[2].draw(screen, self.animations[0].position)        # Show the JUMPING_FRONT_UP_LEFT animation
        elif self.state == State.JUMPING_FRONT_DOWN_RIGHT:          # When it is JUMPING_FRONT_DOWN_RIGHT
            self.animations[1].draw(screen, self.animations[0].position)        # Show the JUMPING_FRONT_DOWN_RIGHT animation
        elif self.state == State.JUMPING_FRONT_DOWN_LEFT:           # When it is JUMPING_FRONT_DOWN_LEFT
            self.animations[2].draw(screen, self.animations[0].position)        # Show the JUMPING_FRONT_DOWN_LEFT animation
        elif self.state == State.JUMPING_BACK_UP_LEFT:              # When it is JUMPING_BACK_UP_LEFT
            self.animations[4].draw(screen, self.animations[0].position)        # Show the JUMPING_BACK_UP_LEFT animation
        elif self.state == State.JUMPING_BACK_UP_RIGHT:             # When it is JUMPING_BACK_UP_RIGHT
            self.animations[3].draw(screen, self.animations[0].position)        # Show the JUMPING_BACK_UP_RIGHT animation
        elif self.state == State.JUMPING_BACK_DOWN_LEFT:            # When it is JUMPING_BACK_DOWN_LEFT
            self.animations[4].draw(screen, self.animations[0].position)        # Show the JUMPING_BACK_DOWN_LEFT animation
        elif self.state == State.JUMPING_BACK_DOWN_RIGHT:           # When it is JUMPING_BACK_DOWN_RIGHT
            self.animations[3].draw(screen, self.animations[0].position)        # Show the JUMPING_BACK_DOWN_RIGHT animation
        else:
            self.animations[0].draw(screen, graphicPosition)        # Show the idle animation
            
    def hit(self):
        '''
        Hits the being.
        @param self The current object
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        self.pleaseRespawn = True      # Request respawn
        self.isMoving = True           # Set moving to true
        self.resetValues()             # Reset the values
        
    def resetValues(self):
        '''
        Resets the values for a new respawn.
        @param self The current object
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        self.jumpMovement.x = 0       # Movement x to 0
        self.movementSlice.x = 0      # Movement slice x to 0
        self.jumpMovement.y = 0       # Movement y to 0
        self.movementSlice.y = 0      # Movement slice y to 0
        
    def dropOff(self):
        '''
        Sets the being to fall definitely.
        @param self The current object
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        self.resetValues()             # Reset the values
        self.isFalling = True          # Set falling to true
        self.landed = False            # Set landed to false
        
    def land(self):
        '''
        Lands the being.
        @param self The current object
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        if self.height <= 0:            # Check if the height is less than 0
            self.height = 0             # Set the height to 0
            self.state = State.IDDLE    # Set the state to IDDLE
        self.resetValues()              # Reset the values
        self.isMoving = False           # Set isMoving to false
        
    def respawn(self):
        '''
        Respawns the being.
        @param self The current object
         
        @author: Dario Urdapilleta
        @version 1.0
        @since 13 nov. 2022
        '''
        self.setHeight(QbertBeingClass.maxHeight)                   # Set the height to the max
        self.state = State.JUMPING_FRONT_DOWN_RIGHT                 # Change the state to     JUMPING_FRONT_DOWN_RIGHT
        self.landed = False                                         # Set landed to false
        self.isMoving = True                                        # Set isMoving to true
        self.isFalling = False                                      # Set is Falling to false
        self.pleaseRespawn = False                                  # Request to respawn
        self.stateTime = 0.5                                        # Set the state time to 0.5

File: QbertCSharp/QbertCSharp/QbertBeing.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using QbertCSharp;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Reflection.Metadata;
using System.Text;
using System.Threading.Tasks;

namespace QbertCSharp
{
    /**
     * This class represents A Qbert being such as the player or enemies.
     * 
     * @author Dario Urdapilleta
     * @version 1.0
     * @since 2022-11-09
     *
     */
    internal class QbertBeing
    {
        public const float maxHeight = 100f;             // The maximum jumping height
        public const float deltaHeight = 4f;             // The height change rate
        public const float deltaMovement = 0.02f;        // The movement animation frame rate
        public const float deltaJump = 0.1f;             // The jump animation frame rate

        public enum State
        {                               // Enum with all the animation states
            IDDLE,
            JUMPING_FRONT_UP_RIGHT,
            JUMPING_FRONT_UP_LEFT,
            JUMPING_FRONT_DOWN_RIGHT,
            JUMPING_FRONT_DOWN_LEFT,
            JUMPING_BACK_UP_LEFT,
            JUMPING_BACK_UP_RIGHT,
            JUMPING_BACK_DOWN_LEFT,
            JUMPING_BACK_DOWN_RIGHT
        }

        private Vector2 position;                               // The being's position
        private Texture2D texture;                              // The being's sprite list
        private Vector2 movementSlice;                          // The direction it is moving
        private Vector2 jumpMovement;                           // The actual placement of the being

        SpriteAnimation[] animations;                           // The animations

        private State state;                                    // The being's current state
        private float stateTime;                                // Time for the key frames
        private float height;                                   // The jump height
        private bool isMoving;                                  // True if the being is moving
        private bool landed;                                    // True if it has landed
        private bool isFalling;                                 // True if it's outside of the board
        protected bool pleaseRespawn;                           // True if it needs to respawn
        /**
         * Creates a new instance of a QbertBeing given the spritesheet information.
         * @param fileName The Spritesheet's file name.
         * @param start The start spritesheet position.
         * @param animationLength The length of each animation.
         * @return A new instance of the QbertBeing
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public QbertBeing(Texture2D sprites, int start, int animationLength)
        {
            this.state = QbertBeing.State.IDDLE;                                                // Start the state with IDDLE
            this.isMoving = false;                                                              // Make the being not move
            this.landed = true;                                                                 // It starts on the ground
            this.isFalling = false;                                                             // It starts in the board
            this.pleaseRespawn = false;                                                         // It doesn't need to be respawned
            this.movementSlice = new Vector2(0, 0);                                             // It's not moving
            this.jumpMovement = new Vector2(0, 0);                                              // It's not moving
            this.texture = sprites;                             					            // Load the sprites
            this.animations = new SpriteAnimation[5];                                           // Create the animations array
            for (int animationCounter = 0; animationCounter < this.animations.Length;animationCounter++)                    // Loop through the animations
            {
                this.animations[animationCounter] = new SpriteAnimation(sprites, 8, 60, 60, start + animationCounter);      // Create the animation
                this.animations[animationCounter].isLooping = true;                                                         // Set it to loop
                this.animations[animationCounter].SetFramesPersecond(8);                                                    // Set the frames per second
            }
            this.stateTime = 0f;                                                                // Reset the state time
            this.height = 0f;																	// Set the height as 0
	    }
	    /**
	        * Deletes all the graphical elements of a being
	        * 
	        * @author Dario Urdapilleta
	        * @version 1.0
	        * @since 2022-11-09
	        *	 
	        */
	    public void Dispose()
        {
            this.texture.Dispose();      // Delete the sprite list
        }
        /**
         * Returns true if the being needs to respawn.
         * @return True if it needs to be respwned.
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public bool NeedsRespawn()
        {
            return this.pleaseRespawn;      // Please respawn this being
        }
        /**
         * Returns true if the being is standing.
         * @return True if the being is standing.
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public bool IsStanding()
        {
            return this.state == QbertBeing.State.IDDLE;   // Make sure the position is not null, and the state is IDDLE
        }
        /**
         * Returns true if the being is moving.
         * @return True if the being is moving.
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public bool IsMoving()
        {
            return this.isMoving;       // Return if the being is moving
        }
        /**
         * Returns true if the being has landed.
         * @return True if the being has landed.
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public bool Landed()
        {
            return this.landed;         // Return true if the bieng has landed
        }
        /**
         * Returns true if the being is falling.
         * @return True if the being is falling.
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public bool IsFalling()
        {
            return this.isFalling;      // Return true if the being is falling
        }
        /**
         * Sets the being's height
         * @param height The being's height.
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public void SetHeight(float height)
        {
            this.height = height;       // Set the height
        }
        /**
         * Returns the being's actual movement
         * @return The being's actual movement
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public Vector2 GetJumpMovement()
        {
            return this.jumpMovement;       // return the actual movement
        }
        /**
         * Returns the being's position as a Vector2
         * @return The being's position as a Vector2
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public Vector2 GetPosition()
        {
            return this.position;       // Return the position
        }
        /**
         * Sets the being's position
         * @param position The being's new position.
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public void SetPosition(Vector2 position)
        {
            this.position = position;   // Set the position
        }
        /**
         * Moves the being torwards a direction.
         * @param direction The direction the being is moving.
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public void Move(int direction)
        {
            this.isMoving = true;                                       // Set isMoving to true
            this.landed = false;                                        // Set landed to false
            switch (direction)                                          // Do differently depending on the direction
            {
                case 0:                                                     // north west
                    this.movementSlice.Y = -QbertBeing.deltaMovement;       // Move negative on y
                    this.movementSlice.X = 0f;                              // Leave x unchanged
                    this.state = QbertBeing.State.JUMPING_BACK_UP_LEFT;     // Change state to JUMPING_BACK_UP_LEFT
                    break;
                case 1:                                                     // north east
                    this.movementSlice.X = -QbertBeing.deltaMovement;       // Move negative on x
                    this.movementSlice.Y = 0f;                              // Leave y unchanged
                    this.state = QbertBeing.State.JUMPING_BACK_UP_RIGHT;    // Change state to JUMPING_BACK_UP_RIGHT
                    break;
                case 2:                                                     // south east
                    this.movementSlice.Y = QbertBeing.deltaMovement;        // Move positive on y
                    this.movementSlice.X = 0f;                              // Leave x unchanged
                    this.state = QbertBeing.State.JUMPING_FRONT_UP_RIGHT;   // Change state to JUMPING_FRONT_UP_RIGHT
                    break;
                default:                                                    // south west
                    this.movementSlice.X = QbertBeing.deltaMovement;        // Move positive on x
                    this.movementSlice.Y = 0f;                              // Leave y unchanged
                    this.state = QbertBeing.State.JUMPING_FRONT_UP_LEFT;    // Change state to JUMPING_FRONT_UP_LEFT
                    break;
            }
            this.stateTime = 0f;                                        // Reset the state time
        }
        /**
         * Updates the being
         * @param gameTime The game time transcurred since last update.
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         */
        public void Update (GameTime gameTime)
        {
            this.stateTime += (float)gameTime.ElapsedGameTime.TotalSeconds;                     // Increase the animation state
            switch (this.state)                                                                 // Do different depending on the state
            {
                case State.IDDLE:                                                               // When it is IDDLE
                    this.height = 0f;                                                           // Set the height to 0
                    break;
                case State.JUMPING_FRONT_UP_RIGHT:                                              // When it is JUMPING_FRONT_UP_RIGHT
                    this.height += QbertBeing.deltaHeight;                                      // Increase the height
                    break;
                case State.JUMPING_FRONT_UP_LEFT:                                               // When it is JUMPING_FRONT_UP_LEFT
                    this.height += QbertBeing.deltaHeight;                                      // Increase the height
                    break;
                case State.JUMPING_FRONT_DOWN_RIGHT:                                            // When it is JUMPING_FRONT_DOWN_RIGHT
                    this.height -= QbertBeing.deltaHeight;                                      // Decrease the height
                    break;
                case State.JUMPING_FRONT_DOWN_LEFT:                                             // When it is JUMPING_FRONT_DOWN_LEFT
                    this.height -= QbertBeing.deltaHeight;                                      // Decrease the height
                    break;
                case State.JUMPING_BACK_UP_LEFT:                                                // When it is JUMPING_BACK_UP_LEFT
                    this.height += QbertBeing.deltaHeight;                                      // Increase the height
                    break;
                case State.JUMPING_BACK_UP_RIGHT:                                               // When it is JUMPING_BACK_UP_RIGHT
                    this.height += QbertBeing.deltaHeight;                                      // Increase the height
                    break;
                case State.JUMPING_BACK_DOWN_LEFT:                                              // When it is IDDLE
                    this.height -= QbertBeing.deltaHeight;                                      // Decrease the height
                    break;
                case State.JUMPING_BACK_DOWN_RIGHT:                                             // When it is IDDLE
                    this.height -= QbertBeing.deltaHeight;                                      // Decrease the height
                    break;
                default:
                    this.height = 0f;                                                           // Set the height to 0
                    break;
            }
            if (this.height >= QbertBeing.maxHeight)                                            // If the being has reached the max height
            {
                switch (this.state)                                                             // Do differently depending on the state
                {
                    case State.JUMPING_FRONT_UP_RIGHT:                                          // JUMPING_FRONT_UP_RIGHT
                        this.state = QbertBeing.State.JUMPING_FRONT_DOWN_RIGHT;                 // Change to JUMPING_FRONT_DOWN_RIGHT
                        break;
                    case State.JUMPING_FRONT_UP_LEFT:                                           // JUMPING_FRONT_UP_LEFT
                        this.state = QbertBeing.State.JUMPING_FRONT_DOWN_LEFT;                  // Change to JUMPING_FRONT_DOWN_LEFT
                        break;
                    case State.JUMPING_BACK_UP_LEFT:                                            // JUMPING_BACK_UP_LEFT
                        this.state = QbertBeing.State.JUMPING_BACK_DOWN_LEFT;                   // Change to JUMPING_BACK_DOWN_LEFT
                        break;
                    case State.JUMPING_BACK_UP_RIGHT:                                           // JUMPING_BACK_UP_RIGHT
                        this.state = QbertBeing.State.JUMPING_BACK_DOWN_RIGHT;                  // Change to JUMPING_BACK_DOWN_RIGHT
                        break;
                    default:                                                                    // Otherwise, do nothing
                        break;
                }
            }
            this.jumpMovement.X += this.movementSlice.X;                                        // Move the being on the x axis
            this.jumpMovement.Y += this.movementSlice.Y;                                        // Move the being to the y axis
            if (!this.isFalling && this.height < 0)                                             // Check if it's not falling and it has landed
            {
                if (this.jumpMovement.X >= 1f)                                                  // Check the movement on x is more than 1
                {
                    this.position.X++;                                                          // Increase the position in x
                }
                if (this.jumpMovement.X <= -1f)                                                 // Check the movement on x is lees than -1
                {
                    this.position.X--;                                                          // Decrease X
                }
                if (this.jumpMovement.Y >= 1f)                                                  // Check the y movement is greater than 1
                {
                    this.position.Y++;                                                          // Increase y
                }
                if (this.jumpMovement.Y <= -1f)                                                 // Check the y movement is less than -1
                {
                    this.position.Y--;                                                          // Decrease y
                }
                this.landed = true;                                                             // Set landed to true
            }
            if (this.height <= -300f)                                                           // Check if the height has gone less than -300
            {
                this.pleaseRespawn = true;                                                      // Request to respawn
            }
            for (int animationCounter = 0; animationCounter < this.animations.Length; animationCounter++)                    // Loop through the animations
            {
                this.animations[animationCounter].Update(gameTime);                             // Update the animation
            }
        }
        /**
         * Draws the being
         * @param spriteBatch The sprite batch that draws it.
         * @param graphicPosition The being's acutal position
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public void Draw(SpriteBatch spriteBatch, Vector2 graphicPosition)
        {
            for (int animationCounter = 0; animationCounter < this.animations.Length; animationCounter++)    // Loop through the animations
            {
                this.animations[animationCounter].position = graphicPosition;                   // Update the animations position
                this.animations[animationCounter].position.Y -= this.height;                    // Add the height to the position

            }
            switch (this.state)                                                                 // Do different depending on the state
            {
                case State.IDDLE:                                                               // When it is IDDLE
                    this.animations[0].Draw(spriteBatch);                                       // Show the idle animation
                    break;
                case State.JUMPING_FRONT_UP_RIGHT:                                              // When it is JUMPING_FRONT_UP_RIGHT
                case State.JUMPING_FRONT_DOWN_RIGHT:                                            // When it is JUMPING_FRONT_DOWN_RIGHT                                                         
                    this.animations[1].Draw(spriteBatch);                                       // Show the jump front right anumation
                    break;
                case State.JUMPING_FRONT_UP_LEFT:                                               // When it is JUMPING_FRONT_UP_LEFT
                case State.JUMPING_FRONT_DOWN_LEFT:                                             // When it is JUMPING_FRONT_DOWN_LEFT
                    this.animations[2].Draw(spriteBatch);                                       // Show the jump front left anumation
                    break;
                case State.JUMPING_BACK_UP_LEFT:                                                // When it is JUMPING_BACK_UP_LEFT
                case State.JUMPING_BACK_DOWN_LEFT:                                              // When it is JUMPING_BACK_DOWN_LEFT
                    this.animations[4].Draw(spriteBatch);                                       // Show the jump front left anumation
                    break;
                case State.JUMPING_BACK_UP_RIGHT:                                               // When it is JUMPING_BACK_UP_RIGHT
                case State.JUMPING_BACK_DOWN_RIGHT:                                             // When it is JUMPING_BACK_DOWN_RIGHT
                    this.animations[3].Draw(spriteBatch);                                       // Show the jump front left anumation
                    break;
                default:
                    this.animations[0].Draw(spriteBatch);                                       // Show the idle anumation
                    break;
            }
        }
        /**
         * Hits the being.
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public void Hit()
        {
            this.pleaseRespawn = true;      // Request respawn
            this.isMoving = true;           // Set moving to true
            this.ResetValues();             // Reset the values
        }
        /**
         * Resets the values for a new respawn.
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public void ResetValues()
        {
            this.jumpMovement.X = 0f;       // Movement x to 0
            this.movementSlice.X = 0f;      // Movement slice x to 0
            this.jumpMovement.Y = 0f;       // Movement y to 0
            this.movementSlice.Y = 0f;      // Movement slice y to 0
        }
        /**
         * Sets the being to fall definitely.
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public void DropOff()
        {
            this.ResetValues();             // Reset the values
            this.isFalling = true;          // Set falling to true
            this.landed = false;            // Set landed to false
        }
        /**
         * Lands the being.
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public void Land()
        {
            if (this.height <= 0)                       // Check if the height is less than 0
            {
                this.height = 0f;                       // Set the height to 0
                this.state = QbertBeing.State.IDDLE;    // Set the state to IDDLE
            }
            this.ResetValues();                         // Reset the values
            this.isMoving = false;                      // Set isMoving to false
        }
        /**
         * Respawns the being.
         * 
         * @author Dario Urdapilleta
         * @version 1.0
         * @since 2022-11-09
         *	 
         */
        public void Respawn()
        {
            this.SetHeight(QbertBeing.maxHeight);                       // Set the height to the max
            this.state = QbertBeing.State.JUMPING_FRONT_DOWN_RIGHT;     // Change the state to	 JUMPING_FRONT_DOWN_RIGHT
            this.landed = false;                                        // Set landed to false
            this.isMoving = true;                                       // Set isMoving to true
            this.isFalling = false;                                     // Set is Falling to false
            this.pleaseRespawn = false;                                 // Request to respawn
            this.stateTime = 0.5f;                                      // Set the state time to 0.5
        }
    }
}
